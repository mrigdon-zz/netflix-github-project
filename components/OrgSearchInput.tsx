import React, { ChangeEvent } from "react";
import styles from "../styles/OrgSearchInput.module.css";
import debounce from "../utils/debounce";
import { label, LocaleContext } from "../utils/i18n";

const examples = [
  "netflix",
  "facebook",
  "linkedin",
  "apple",
  "google",
  "microsoft",
  "salesforce",
  "stripe",
  "paypal",
  "airbnb",
];

export default class OrgSearchInput extends React.Component<{
  loading?: boolean;
  onSearch(text: string): void;
}> {
  static contextType = LocaleContext;

  state = { exampleIndex: 0, searchText: "" };

  private intervalId?: NodeJS.Timer;

  private get placeholder() {
    return label(this.context.locale, "searchForOrgs", {
      example: examples[this.state.exampleIndex],
      className: styles.example,
    });
  }

  componentDidMount() {
    this.setExampleInterval();
  }

  componentWillUnmount() {
    if (this.intervalId) clearInterval(this.intervalId);
  }

  render() {
    const { placeholder } = this;
    return (
      <div className={styles.inputContainer}>
        <input
          aria-label={placeholder}
          className={styles.input}
          value={this.state.searchText}
          onChange={this.handleChangeSearch}
        />

        <span
          className={styles.placeholder}
          dangerouslySetInnerHTML={{ __html: placeholder }}
        ></span>

        {this.props.loading && <span className={styles.spinner}>⚙︎</span>}
      </div>
    );
  }

  private setExampleInterval() {
    this.intervalId = setInterval(() => {
      let newIndex = this.state.exampleIndex + 1;
      if (newIndex === examples.length) newIndex = 0;
      this.setState({ exampleIndex: newIndex });
    }, 3000);
  }

  private handleChangeSearch = (event: ChangeEvent<HTMLInputElement>) => {
    this.setState({ searchText: event.target.value });
    this.search(event.target.value);
  };

  private search = debounce(async (text: string) => {
    this.props.onSearch(text);
  }, 1000);
}
