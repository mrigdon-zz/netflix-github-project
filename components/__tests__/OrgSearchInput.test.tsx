import React from "react";
import TestRenderer from "react-test-renderer";
import OrgSearchInput from "../OrgSearchInput";

describe("OrgSearchInput", () => {
  it("renders the input with a new example every 3 seconds", () => {
    jest.useFakeTimers();

    const output = TestRenderer.create(
      <OrgSearchInput search="" onSearch={jest.fn()} />
    );
    expect(output).toMatchSnapshot("before 3 seconds");

    jest.advanceTimersByTime(3000);
    expect(output).toMatchSnapshot("after 3 seconds");
  });
});
