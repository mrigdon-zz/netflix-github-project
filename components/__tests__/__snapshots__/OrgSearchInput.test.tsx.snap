// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrgSearchInput renders the input with a new example every 3 seconds: after 3 seconds 1`] = `
<div
  className="inputContainer"
>
  <input
    className="input"
    onChange={[Function]}
    value=""
  />
  <span
    className="placeholder"
    dangerouslySetInnerHTML={
      Object {
        "__html": "🔍 Enter a Github org name like <span class=\\"example\\">facebook</span>",
      }
    }
  />
</div>
`;

exports[`OrgSearchInput renders the input with a new example every 3 seconds: before 3 seconds 1`] = `
<div
  className="inputContainer"
>
  <input
    className="input"
    onChange={[Function]}
    value=""
  />
  <span
    className="placeholder"
    dangerouslySetInnerHTML={
      Object {
        "__html": "🔍 Enter a Github org name like <span class=\\"example\\">netflix</span>",
      }
    }
  />
</div>
`;
